JSON Transformation logic

davilimap opened the pull request:
Created all logic for JSON transformations

Adding unit tests for correct JSON transformations

Updating NuGet packages

davilimap committed:
Adding JSON library
Adding transform class

davilimap committed:
Adding trasformation class

davilimap committed:
Adding class for document to transform
Initial default transform logic

davilimap committed:
Finishing default transform logic
Default transformation tests

davilimap committed:
Add object merge test

davilimap committed:
Streamline tests

davilimap committed:
Remove usings
Move test inputs

davilimap committed:
Adding array and object replace tests

davilimap committed:
Adding additional primitive test logic
Adding add aray test
Adding clone to constructor

davilimap committed:
Extracted load to utilities
Added test class for created classes

davilimap committed:
Removing transformer class to be added later

davilimap committed:
Adding verb logic
Adding tests

davilimap committed:
Add property enum
Adding exception
Initial remove logic
Generalize tests
Add initial remove tests

davilimap committed:
Migrate to chain of responsibility pattern
Remove transformation and tests
Adding classes for other transformations

davilimap committed:
Rename logic
Rename tests

davilimap committed:
Adding replace
Adding replace tests
Altering test names for readability
Adding skipped tests to be implemented

davilimap committed:
Adding merge logic

davilimap committed:
Adding build time transformation logic

davilimap committed:
Saved document should be formatted correctly
Adding preview and add transforms for JSON

davilimap committed:
Revert "Saved document should be formatted correctly"

This reverts commit 1909a675da33f8d4320a81a62f7c39aeb3797fa5.

davilimap committed:
Revert "Adding build time transformation logic"

This reverts commit 4e943cdebddb7c426230631c4f2d7b57126b0a7b.

davilimap committed:
Remove deleted file from project

davilimap committed:
Adding merge to transformation chain
Adding primitive tests for merge

davilimap committed:
Adding merge object tests

davilimap committed:
Merge branch 'upstream-master' into verb-transformations

davilimap committed:
Adding new lines

davilimap committed:
Adding skipped test

davilimap committed:
Adding utility tests

davilimap committed:
Move JDT to new project
Update nuget packages
Update VisualStudio.Tests output

davilimap committed:
Remove unused class
Remove unnecessary nuget reference

davilimap committed:
Remove usings

jviau commented on SlowCheetah.JDT/JsonUtilities.cs:
```diff
@@ -0,0 +1,120 @@
+                {
+                    CommentHandling = CommentHandling.Ignore,
+
+                    // Obs: LineInfo is handled on Ignore and not Load
```
Include a link to the newtonsoft GitHub issue you opened for this

davilimap replied:
This is has been fixed, just need to wait for the next release of Newtonsoft to update the NuGet package. I'll still link to the issue though.

jviau commented on SlowCheetah.JDT/JsonUtilities.cs:
```diff
@@ -0,0 +1,120 @@
+        /// <returns>The string property. Null if the property does is not JDT syntax</returns>
+        internal static string GetJdtSyntax(string key)
+        {
+            if (!IsJdtSyntax(key))
```
This can be collapsed into a single line: `return IsJdtSyntax(key) ? key.SubString(JdtSyntaxPrefix.Length) : null;`

jviau commented on SlowCheetah.JDT/JsonUtilities.cs:
```diff
@@ -0,0 +1,120 @@
+        /// <param name="node">The node to base the search on</param>
+        /// <param name="path">The JSONPath to find nodes.</param>
+        /// <returns>The corresponding tokens</returns>
+        internal static IEnumerable<JToken> GetTokensFromPath(JObject node, string path)
```
Maybe make this a an extension method? `this JObject node`

jviau commented on SlowCheetah.JDT/JsonUtilities.cs:
```diff
@@ -0,0 +1,120 @@
+        /// </summary>
+        /// <param name="token">The token to verify</param>
+        /// <param name="errorMessage">Error message</param>
+        internal static void ThrowIfRoot(JToken token, string errorMessage)
```
Could also be an extension method.

jviau commented on SlowCheetah.JDT/JsonUtilities.cs:
```diff
@@ -0,0 +1,120 @@
+        /// </summary>
+        /// <param name="original">Array to merge into</param>
+        /// <param name="arrayToMerge">Array to be merged</param>
+        internal static void MergeArray(JArray original, JArray arrayToMerge)
```
This could be an extension method.

jviau replied:
arg validation.

jviau commented on SlowCheetah.JDT/Processors/JdtDefault.cs:
```diff
@@ -0,0 +1,51 @@
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// Represents a recursive JDT transformation
```
Update comment.

jviau commented on SlowCheetah.JDT/Processors/JdtDefault.cs:
```diff
@@ -0,0 +1,51 @@
+                    }
+                    else if (nodeToTransform.Type != JTokenType.Object || transformNode.Value.Type != JTokenType.Object)
+                    {
+                        // TO DO: Verify if object has JDT verbs
```
What are you planning on verifying here? What happens if it does have verbs?

davilimap replied:
We should possibly generate an error if the default transformation (and the other transformations as well) results in an addition or replacement of a node with an object that contains verbs. Those verbs are being processed because we don't recurse into the node.

jviau commented on SlowCheetah.JDT/Processors/JdtMerge.cs:
```diff
@@ -0,0 +1,113 @@
+    /// </summary>
+    internal class JdtMerge : JdtProcessor
+    {
+        private const string PathAttribute = "path";
```
These are used by several jdt verbs - move them to a common location.

jviau commented on SlowCheetah.JDT/Processors/JdtMerge.cs:
```diff
@@ -0,0 +1,113 @@
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// Represents a recursive JDT transformation
```
Update comment.

jviau commented on SlowCheetah.JDT/Processors/JdtMerge.cs:
```diff
@@ -0,0 +1,113 @@
+            }
+            else if (hasPath && hasValue)
+            {
+                if (mergeObject.Properties().Where(p => !p.Name.Equals(pathFullAttribute) && !p.Name.Equals(valueFullAttribute)).Count() > 0)
```
Change from `.Where` to `.Any`. Or switch to a `TryGetValue || TryGetValue`

jviau commented on SlowCheetah.JDT/JsonDocument.cs:
```diff
@@ -0,0 +1,72 @@
+        /// <param name="filePath">Path to the JSON file</param>
+        public JsonDocument(string filePath)
+        {
+            this.documentPath = filePath;
```
arg validation

jviau commented on SlowCheetah.JDT/JsonDocument.cs:
```diff
@@ -0,0 +1,72 @@
+        /// <param name="docObject">The object to be transformed</param>
+        public JsonDocument(JObject docObject)
+        {
+            this.documentPath = string.Empty;
```
arg validation. As discussed in person, we should rethink having newtonsoft.json types exposed in this API.

jviau commented on SlowCheetah.JDT/JsonDocument.cs:
```diff
@@ -0,0 +1,72 @@
+        /// <inheritdoc/>
+        public bool Equals(JsonDocument other)
+        {
+            return JToken.DeepEquals(this.documentObject, other.documentObject);
```
Arg validation.

jviau commented on SlowCheetah.JDT/JsonTransformation.cs:
```diff
@@ -0,0 +1,48 @@
+        /// <param name="transformFile">File that defines the trasnformation</param>
+        public JsonTransformation(string transformFile)
+        {
+            this.transform = JsonUtilities.LoadObjectFromFile(transformFile);
```
arg validation.

jviau commented on SlowCheetah.JDT/JsonTransformation.cs:
```diff
@@ -0,0 +1,48 @@
+        /// <param name="document">Document to be transformed</param>
+        public void Apply(JsonDocument document)
+        {
+            JdtProcessor.ProcessTransform(document.GetObject(), (JObject)this.transform.DeepClone());
```
arg validation.

jviau commented on SlowCheetah.JDT/Processors/JdtMerge.cs:
```diff
@@ -0,0 +1,113 @@
+            this.Successor.Process(source, transform);
+        }
+
+        private void Merge(JObject source, JToken mergeValue, bool allowArray)
```
I am not sure I like `bool allowArray`. You could have an outer `Merge` which only checks for array case, then a `MergeCore` which is the equivalent to `allowArray = false`

jviau commented on SlowCheetah.JDT/Processors/JdtMerge.cs:
```diff
@@ -0,0 +1,113 @@
+        private void MergeWithObject(JObject source, JObject mergeObject)
+        {
+            JToken pathToken, valueToken;
+            string pathFullAttribute = JsonUtilities.JdtSyntaxPrefix + PathAttribute;
```
Move to a `const` in a utilities class?

jviau commented on SlowCheetah.JDT/Processors/JdtMerge.cs:
```diff
@@ -0,0 +1,113 @@
+                }
+
+                var tokensToMerge = JsonUtilities.GetTokensFromPath(source, pathToken.ToString());
+                foreach (JToken tokenToMerge in tokensToMerge.ToList())
```
why `ToList()`?

davilimap replied:
`IEnumerable` does not support altering the contents. I'll just have the utility method return the list

jviau commented on SlowCheetah.JDT/Processors/JdtProcessor.ProcessorChain.cs:
```diff
@@ -0,0 +1,98 @@
+    {
+        private class JdtProcessorChain
+        {
+            private readonly List<JdtProcessor> processors = new List<JdtProcessor>()
```
Add a comment that this is the ordered list of processors we currently support.

jviau commented on SlowCheetah.JDT/Processors/JdtProcessor.ProcessorChain.cs:
```diff
@@ -0,0 +1,98 @@
+            {
+                var validator = this.processors.First() as JdtValidator;
+
+                // The successor if each transform processor should be the next one on the list
```
spelling: if -> of

davilimap committed:
Change document object to property

jviau commented on SlowCheetah.JDT/Processors/JdtProcessor.ProcessorChain.cs:
```diff
@@ -0,0 +1,98 @@
+            }
+        }
+
+        private class JdtValidator : JdtProcessor
```
Move this to its own file. It isn't chain specific, like the `JdtEndOfChain`, but more akin to the other processors.

jviau commented on SlowCheetah.JDT/Processors/JdtRecurse.cs:
```diff
@@ -0,0 +1,44 @@
+        public override void Process(JObject source, JObject transform)
+        {
+            // Nodes that should be removed from the transform after they are handled
+            List<string> nodesToRemove = new List<string>();
```
nit: use `var`

jviau commented on SlowCheetah.JDT/Processors/JdtRecurse.cs:
```diff
@@ -0,0 +1,44 @@
+            }
+
+            // Remove all of the previously handled nodes
+            nodesToRemove.ForEach(node => transform.Remove(node));
```
mutating the transform object? Can you explain why?

davilimap replied:
Because each transformation is containerized, later transformations need a way to know if a node has already been processed in case of a removal or rename in the source object. 
The transform object is not actually mutated as a deep clone is passed in, per [this](https://github.com/davilimap/slow-cheetah/pull/6/files#diff-5d0dbb0e3b7e365048b4495556f01a22R45).

jviau replied:
I see. It does go against the pattern we established, where 2 processes can work with the same node. But I see how _Recurse_ can be considered a special case here.

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,167 @@
+    {
+        private const string PathAttribute = "path";
+
+        private bool removedThisNode;
```
The current processor chain shares one instance of `JdtRemove`. You cannot have state like this, it will corrupt other transforms.

Remove this field and have the instance functions return a bool which performs the same function.

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,167 @@
+            this.removedThisNode = false;
+
+            JToken removeValue;
+            if (transform.TryGetValue(JsonUtilities.JdtSyntaxPrefix + this.Verb, out removeValue))
```
Consider a `TryGetJdtValue` extension method which auto-concats the verb with the prefix.

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,167 @@
+                case JTokenType.String:
+                    // TO DO: warning if unable to remove
+                    // If the value is just a string, remove that node
+                    source.Remove(removeValue.ToObject<string>());
```
why `ToObject<string>()`? I think I saw you using `ToString()` in other places. Can you use that here?

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,167 @@
+            JToken removeValue;
+            if (transform.TryGetValue(JsonUtilities.JdtSyntaxPrefix + this.Verb, out removeValue))
+            {
+                this.Remove(source, removeValue, true);
```
`this.Remove` should return true/false indicating to continue or not. Then this can change to `if (!this.Remove(source, removeValue, true) { return; }`. Then the `this.Successor.Process` does not need to be conditional below.

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,167 @@
+            }
+        }
+
+        private void Remove(JObject source, JToken removeValue, bool allowArray)
```
The `allowArray` special case appears to be common among a few processors. Consider extracting to a base class to consolidate the iteration when the transform argument is an array. If it is possible that is. Other issues may prevent that.

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,167 @@
+                    source.Remove(removeValue.ToObject<string>());
+                    break;
+                case JTokenType.Boolean:
+                    if (removeValue.ToObject<bool>())
```
`ToObject` is superfluous here. You already checked the type, you can cast it directly.
If we update appveyor and use C#7 you may be able to improve this even more with pattern matching.

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,167 @@
+        private void RemoveWithAttributes(JObject source, JObject removeObject)
+        {
+            string pathFullAttribute = JsonUtilities.JdtSyntaxPrefix + PathAttribute;
+            if (removeObject.Properties().Where(p => !p.Name.Equals(pathFullAttribute)).Count() > 0)
```
`TryGetValue` You are only looking for  1 property.

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,167 @@
+
+        private void RemoveThisNode(JObject nodeToRemove)
+        {
+            // If the value is true, everything from the node and set it to null
```
Fix this comment -> "remove everything"

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,167 @@
+        private void RemoveThisNode(JObject nodeToRemove)
+        {
+            // If the value is true, everything from the node and set it to null
+            if (nodeToRemove.Root.Equals(nodeToRemove))
```
Isn't this the same as your utility function to check if it is the root?

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,167 @@
+                throw new JdtException("You cannot remove the root");
+            }
+
+            JProperty parent = (JProperty)nodeToRemove.Parent;
```
nit: user `var` when the type is apparent, like a cast or a `new`

jviau commented on SlowCheetah.JDT/Processors/JdtRename.cs:
```diff
@@ -0,0 +1,130 @@
+        private void RenameWithProperties(JObject source, JObject renameProperties)
+        {
+            JToken pathToken, valueToken;
+            string pathFullAttribute = JsonUtilities.JdtSyntaxPrefix + PathAttribute;
```
move to a const on a utilities class. Or a const on `JdtProcessor`

jviau commented on SlowCheetah.JDT/Processors/JdtReplace.cs:
```diff
@@ -0,0 +1,126 @@
+        private const string PathAttribute = "path";
+        private const string ValueAttribute = "value";
+
+        private bool replacedThisNode;
```
Similar to `Remove`, you cannot have instance fields in the current design.

jviau reviewed (changes requested):
I am seeing a lot of duplicate code in the various `JdtProcessor` implementations. Look them over and extract these common parts to base classes.

For example, the `jdt.path` and `jdt.value` strings  can be extracted. Also you could have the conditional check extracted and have derived classes say which combination of the 2 they support.

The `allowArray` design can be extracted to a `bool ProcessInternal` and `bool ProcessCore` style (names up to you to choose). Where one level checks for the array loop, and the next level assumes no array loop. And then the return value indicates to continue to `Successor` or not.

davilimap committed:
Remove allowArray from transforms
Removed instance fields for imutable classes
Extracted JdtValidator to own class
Other PR changes

AArnott commented on SlowCheetah.JDT/JsonUtilities.cs:
```diff
@@ -0,0 +1,123 @@
+    internal static class JsonUtilities
+    {
+        /// <summary>
+        /// The suffix for all JDT syntax
```
I think you mean the _prefix_ for all JDT syntax.

AArnott commented on SlowCheetah.JDT/Processors/JdtRename.cs:
```diff
@@ -0,0 +1,127 @@
+        public override void Process(JObject source, JObject transform)
+        {
+            JToken renameValue;
+            if (transform.TryGetValue(JsonUtilities.JdtSyntaxPrefix + this.Verb, out renameValue))
```
Avoid dereferencing an argument without checking for null first.
I'll just say that here, but I see several violations of this rule in this review.

AArnott commented on SlowCheetah.JDT/Processors/JdtRename.cs:
```diff
@@ -0,0 +1,127 @@
+                    {
+                        if (renameOperation.Value.Type != JTokenType.String)
+                        {
+                            throw new JdtException("Rename value must be a string");
```
If this string will appear to the user, I suspect it should be localized.

AArnott commented on SlowCheetah/Properties/AssemblyInfo.cs:
```diff
@@ -25,6 +25,9 @@
 [assembly: Guid("6354d859-e629-49fc-b154-fc0ba42d71b0")]
 
+// Allow the test to see the internals of this class
+[assembly: InternalsVisibleTo("SlowCheetah.Tests")]
```
I advise against InternalsVisibleTo. By testing only the public API, you are free to refactor all you want without breaking any tests. That tends to mean more refactoring (due to reduced costs and no anxiety about breaking tests), cleaner code, and fewer bugs! :smile:
It also forces you to realize when writing tests that you're missing public API required to do what you want (whether that's provide mocks, assert things, etc.) and you end up building a better API for everyone. Chances are if you test needs to do something, some consumer of your library will too.

davilimap replied:
That is a very good point. Currently the only reason the internals are visible to the tests are for testing some utility methods. I don't think these methods should be public, but at the same time, I think it's good to test them so they don't hinder the API tests.

AArnott commented on SlowCheetah/SlowCheetah.csproj:
```diff
@@ -86,7 +86,9 @@
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
-    <None Include="packages.config" />
+    <None Include="packages.config">
```
Consider the newer world of project.json or the much better and latest world of PackageReference instead of packages.config. It requires VS 2017, but hey, we shipped it. :)

jviau replied:
Moving to .netcore with package references is something we will do eventually.

AArnott commented on SlowCheetah.JDT.Tests/TransformTest.cs:
```diff
@@ -0,0 +1,179 @@
+        [MemberData(nameof(GetRemoveInputs))]
+        public void Remove(string testFileName)
+        {
+            BaseTransformTest(RemoveTestDirectory, testFileName);
```
I haven't deeply studied your unit tests, but it at least appears that your tests are super small and concise. That's *great* as that tends to lead to more tests being written now and in the future when folks add functionality or fix bugs and ought to add regression tests. Which of course leads to higher code quality and fewer bugs.

davilimap committed:
Adding validation
Moving utility methods to extension methods
Other PR changes

davilimap committed:
Adding abstraction for processors that handle arrays

davilimap committed:
Remove dependant object constructor
Remove usings

davilimap committed:
Migrate to streams instead of classes
Updating tests to use stream

davilimap committed:
Add comment

davilimap committed:
Adding argument verification

jviau commented on SlowCheetah.JDT/Processors/JdtArrayProcessor.cs:
```diff
@@ -0,0 +1,80 @@
+        /// <param name="source">Object to be transformed</param>
+        /// <param name="transformValue">Value of the transform</param>
+        /// <returns>True if transforms should be halted</returns>
+        protected bool Transform(JObject source, JToken transformValue)
```
Why `protected`? Make this private until you find a use for actually overriding it.

jviau commented on SlowCheetah.JDT/Processors/JdtArrayProcessor.cs:
```diff
@@ -0,0 +1,80 @@
+        /// <param name="source">Object to be transformed</param>
+        /// <param name="transformValue">Value of the transform</param>
+        /// <returns>True if transforms should be halted</returns>
+        protected abstract bool TransformCore(JObject source, JToken transformValue);
```
nit: I prefer `ProcessCore`
Also, `true` to halt feels weird to me. I prefer `true` to continue processing.

jviau commented on SlowCheetah.JDT/JsonDocument.cs:
```diff
@@ -0,0 +1,62 @@
+        /// <summary>
+        /// Gets the JObject corresponding to the root of the document
+        /// </summary>
+        internal JObject DocumentObject { get; private set; }
```
I see no reason for `internal` here - `public` is better. The user can apply the transform, but the only way to get the result is to save it to a file and then re-read it from disk?

davilimap replied:
Removing this. See #7

jviau commented on SlowCheetah.JDT/JsonUtilities.cs:
```diff
@@ -0,0 +1,99 @@
+    internal static class JsonUtilities
+    {
+        /// <summary>
+        /// The preffix for all JDT syntax
```
spelling: "prefix"

jviau commented on SlowCheetah.JDT/JsonUtilities.cs:
```diff
@@ -0,0 +1,99 @@
+        /// <param name="node">The node to base the search on</param>
+        /// <param name="path">The JSONPath to find nodes.</param>
+        /// <returns>The corresponding tokens</returns>
+        internal static List<JToken> GetTokensFromPath(JObject node, string path)
```
Why does this return a `List<T>`? Your only uses of it are in `foreach` statements. `IEnumerable<T>` would be preferred. Even then all this function does is save yourself passing in a bool. How useful is it to begin with?

davilimap replied:
1. It needs to be a list because the original collection is altered, which makes `IEnumerable` unusable.
2. The original reason this was a utility method was because the JSONPath functionality was going to be extended (the original plan was to support relative and absolute paths). I'll just remove it and if that functionality is extended, I'll just refactor the code.

jviau commented on SlowCheetah.JDT/JsonUtilities.cs:
```diff
@@ -0,0 +1,99 @@
+        /// <param name="token">The token to verify</param>
+        /// <param name="errorMessage">Error message</param>
+        internal static void ThrowIfRoot(JToken token, string errorMessage)
+        {
```
Arg validation.

jviau commented on SlowCheetah.JDT/JsonDocument.cs:
```diff
@@ -0,0 +1,62 @@
+    /// <summary>
+    /// Represents a JSON file to be transformed
+    /// </summary>
+    public class JsonDocument : IEquatable<JsonDocument>
```
`IEquatable<JsonDocument>` has no use outside of testing. Even then the implementation is very test-specific. Make the JObject publicly accessible and remove `IEquatable`.

It's implementation isn't even corrent: 2 JsonDocuments can be the JObject, but different file paths and they will be equal. That is incorrect.

davilimap replied:
`JsonDocument` will be removed

jviau commented on SlowCheetah.JDT/JsonTransformation.cs:
```diff
@@ -0,0 +1,44 @@
+    /// <summary>
+    /// Represents a transformation based on a JSON file using JDT
+    /// </summary>
+    public class JsonTransformation
```
I am not really sold on the usefulness of `JsonDocument` and `JsonTransformation`. All they are is adding the file on disk information with the `JObject`. I do not think JDT should be responsible for reading/writing JSON from/to disk - leave that to the consumer of this API to decide how to do that. We can either work directly with `JObject` or if we want to not have newtonsoft in our public API then we can operate on streams.

As for eventually needing the file name during logging, this can be solved another way. Perhaps an overload for `JdtProcessor.ProcessTransform` with a _logging context_ style object that contains the names of the source and transform files.

davilimap replied:
JsonTransformation now operates with streams of files directly, always returning a stream for the user to save

jviau commented on SlowCheetah.JDT/Processors/JdtProcessor.cs:
```diff
@@ -0,0 +1,69 @@
+        /// <summary>
+        /// Executes the entire transformation with the given objects
+        /// </summary>
+        /// <param name="source">Object to be transformed</param>
```
This should document that it mutates `source`

jviau commented on SlowCheetah.JDT/JsonTransformation.cs:
```diff
@@ -0,0 +1,44 @@
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            JdtProcessor.ProcessTransform(document.DocumentObject, (JObject)this.transform.DeepClone());
```
the `transform.DeepClone()` should be performed inside `JdtProcessor.ProcessTransform`. Not here.

davilimap committed:
Changing return of JdtArrayProcessor

lifengl commented on SlowCheetah.JDT/JdtExtensions.cs:
```diff
@@ -0,0 +1,51 @@
+        /// </summary>
+        /// <param name="original">Array to merge into</param>
+        /// <param name="arrayToMerge">Array to be merged</param>
+        internal static void MergeInto(this JArray original, JArray arrayToMerge)
```
The name of this function is somewhat confusing. A.MergeInto(B) actually merge B into A. Maybe we should not use 'Into' here?

davilimap replied:
Turns out this method is useless. Removed it

davilimap committed:
Change constructor of transformation
Fix tests

davilimap committed:
Removing unnecessary utility method
Adding argument validation

davilimap committed:
Replaced extension method with appropriate native method
Adding documentation

davilimap committed:
Merge branch 'verb-transformations' into streams

davilimap committed:
Merge pull request #7 from davilimap/streams

Migrating to streams

davilimap committed:
Creating validator
Adding validator to merge, remove and rename

davilimap committed:
Changing validator logic to accomodate wrong jdt syntax

davilimap committed:
Adding validator to merge

davilimap committed:
Adding comments
Removing headers

davilimap committed:
Remove headers

davilimap committed:
Merge branch 'validator' into verb-transformations

davilimap committed:
Adding comments
Removing header

jviau commented on SlowCheetah.JDT/Processors/Attributes/AttributeExtensions.cs:
```diff
@@ -0,0 +1,56 @@
+    /// <summary>
+    /// Implements extensions for <see cref="JdtAttributes"/>
+    /// </summary>
+    internal static class AttributeExtensions
```
Can you rename this to `JdtAttributeExtensions` - just so it is not confused with general C# attributes.

jviau commented on SlowCheetah.JDT/Processors/Attributes/JdtAttributeValidator.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <param name="validAttributes">The attributes that are valid</param>
+        internal JdtAttributeValidator(params JdtAttributes[] validAttributes)
+        {
+            this.validAttributes = validAttributes.ToList();
```
Argument validation - it is still possible for `params` style arguments to be null.

jviau commented on SlowCheetah.JDT/Processors/Attributes/JdtAttributeValidator.cs:
```diff
@@ -0,0 +1,85 @@
+        /// An empty dictionary is returned if no valid properties are found</returns>
+        internal Dictionary<JdtAttributes, JToken> ValidateAndReturnAttributes(JObject transformObject)
+        {
+            Dictionary<JdtAttributes, JToken> attributes = new Dictionary<JdtAttributes, JToken>();
```
Argument validation.

jviau commented on SlowCheetah.JDT/Processors/JdtMerge.cs:
```diff
@@ -0,0 +1,98 @@
+    /// </summary>
+    internal class JdtMerge : JdtArrayProcessor
+    {
+        private JdtAttributeValidator attributeValidator;
```
`readonly`

jviau commented on SlowCheetah.JDT/Processors/Attributes/JdtAttributeValidator.cs:
```diff
@@ -0,0 +1,85 @@
+    /// </summary>
+    internal class JdtAttributeValidator
+    {
+        private List<JdtAttributes> validAttributes;
```
`readonly` Also consider `IReadonlyCollection<T>`

jviau commented on SlowCheetah.JDT/Processors/JdtRemove.cs:
```diff
@@ -0,0 +1,127 @@
+    /// </summary>
+    internal class JdtRemove : JdtArrayProcessor
+    {
+        private JdtAttributeValidator attributeValidator;
```
`readonly`

jviau commented on SlowCheetah.JDT/Processors/JdtRename.cs:
```diff
@@ -0,0 +1,108 @@
+    /// </summary>
+    internal class JdtRename : JdtArrayProcessor
+    {
+        private JdtAttributeValidator attributeValidator;
```
`readonly`

jviau commented on SlowCheetah.JDT/Processors/JdtReplace.cs:
```diff
@@ -0,0 +1,96 @@
+    /// </summary>
+    internal class JdtReplace : JdtArrayProcessor
+    {
+        private JdtAttributeValidator attributeValidator;
```
`readonly`

jviau reviewed (changes requested):
Some more argument validation and and `readonly` opportunities.

davilimap committed:
Argument validation
Other PR changes

davilimap committed:
Removed unused method
